File: LAB_3/LAB_3_Java/src/Storage.java
import java.util.ArrayList;
import java.util.concurrent.Semaphore;

public class Storage {

    private Semaphore access;
    private Semaphore full;
    private Semaphore empty;

    private ArrayList<String> _storage = new ArrayList<>();

    public Storage(int size)
    {
        access= new Semaphore(1);
        full = new Semaphore(size);
        empty = new Semaphore(0);
    }

    public String getItem()
    {
        return _storage.get(0);
    }

    public void addItem(String item)
    {
        _storage.add(item);
    }

    public void removeItem()
    {
        _storage.remove(0);
    }

    public void acquireAccess() throws InterruptedException
    {
        access.acquire();
    }

    public void releaseAccess()
    {
        access.release();
    }

    public void acquireEmpty() throws InterruptedException
    {
        empty.acquire();
    }

    public void releaseEmpty()
    {
        empty.release();
    }

    public void acquireFull() throws InterruptedException
    {
        full.acquire();
    }

    public void releaseFull()
    {
        full.release();
    }
}

File: LAB_3/LAB_3_Csharp/LAB_3_Csharp/Storage.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading;

namespace LAB_3_Csharp
{
    internal class Storage
    {
        public Semaphore access;
        public Semaphore full;
        public Semaphore empty;

        private List<string> _storage = new List<string>();
        private int _storageSize;

        public Storage(int storageSize)
        {
            access = new Semaphore(1, 1);
            full = new Semaphore(storageSize, storageSize);
            empty = new Semaphore(0, storageSize);
            _storageSize = storageSize;
        }

        public string GetItem()
        {
            return _storage.ElementAt(0);
        }

        public void AddItem(string item)
        {
            _storage.Add(item);
        }

        public void RemoveItem()
        {
            _storage.RemoveAt(0);
        }

        public void AcquireAccess()
        {
            access.WaitOne();
        }

        public void ReleaseAccess()
        {
            access.Release();
        }

        public void AcquireEmpty()
        {
            empty.WaitOne();

        }

        public void ReleaseEmpty()
        {
            empty.Release();
        }

        public void AcquireFull()
        {
            full.WaitOne();

        }

        public void ReleaseFull()
        {
            full.Release();
        }
    }
}
